integration/gcpkms/src/default_sa.rs

```diff
@@ -0,0 +1,248 @@
+}
+
+/// Retrieve a specified piece of metadata from the metadata server.
+async fn get_gce_metadata(name: &str) -> Result<String, TinkError> {
+    // Allow the metadata server location to be overridden by environment (to allow testing).
+    let host = std::env::var(METADATA_HOST_ENV).unwrap_or_else(|_e| METADATA_IP_STR.to_string());
+    let authority: http::uri::Authority = host
+        .parse()
+        .map_err(|e| wrap_err("failed to parse host", e))?;
+    let uri = hyper::Uri::builder()
+        .scheme("http")
+        .authority(authority)
+        .path_and_query(format!("/computeMetadata/v1/{}", name))
+        .build()
+        .map_err(|e| wrap_err("failed to build Uri", e))?;
+    let client = hyper::Client::new();
+
+    let req = hyper::Request::builder()
+        .method(http::method::Method::GET)
+        .uri(uri)
+        .header(http::header::USER_AGENT, USER_AGENT)
+        .header("Metadata-Flavor", "Google")
+        .body(hyper::Body::empty())
+        .map_err(|e| wrap_err("failed to build request", e))?;
+    let rsp = client
+        .request(req)
+        .await
+        .map_err(|e| wrap_err("failed to execute request", e))?;
+    if rsp.status() != http::StatusCode::OK {
+        return Err("failed HTTP request".into());
+    }
+    let bytes = hyper::body::to_bytes(rsp.into_body())
+        .await
+        .map_err(|e| wrap_err("failed to retrieve response body", e))?;
+    String::from_utf8(bytes.to_vec()).map_err(|e| wrap_err("failed to convert body to string", e))
+}
+
+/// Token structure for JSON returned by metadata server
+#[derive(Deserialize)]
+struct Token {
+    pub access_token: String,
+    pub expires_in: i64,
+    // Also has `token_type: String` which we ignore.
+}
+
+/// Local copy of [`yup_oauth2::AccessToken`].
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Deserialize, Serialize)]
+struct AccessTokenClone {
+    pub value: String,
+    pub expires_at: Option<DateTime<Utc>>,
+}
+
+pub struct DefaultServiceAccountAuthenticator {
+    // Map from scopelist to access tokens.
+    // We don't attempt to canonicalize the scopelist (so a,b is different than b,a).
+    tokens: RefCell<HashMap<String, yup_oauth2::AccessToken>>,
+}
+
+impl DefaultServiceAccountAuthenticator {
+    pub async fn new() -> Result<Self, TinkError> {
+        if !on_gce().await {
+            return Err("not running on GCE".into());
+        }
+        Ok(Self {
+            tokens: RefCell::new(HashMap::new()),
+        })
+    }
+
+    pub async fn token(&self, scopes: &[&str]) -> Result<yup_oauth2::AccessToken, TinkError> {
+        let scopelist = scopes.join(",");
+
+        if let Some(token) = self.tokens.borrow().get(&scopelist) {
+            // We have a cached token for this scopelist; check its expiry.
+            if let Some(expires_at) = token.expiration_time() {
+                if Utc::now() < expires_at {
+                    return Ok(token.clone());
+                }
+            } else {
```

Comment:
I might have missed it in the code, but there does not seem a case where the expiration time is None. Does this need to be an Option?

Reply:
We don't need it to be an `Option`, but we do need to keep `AccessTokenClone` synced to [`yup_oauth2::AccessToken`](https://docs.rs/yup-oauth2/5.0.3/src/yup_oauth2/types.rs.html#9-12) (which uses an `Option`) to allow the serialize-deserialize hack. 

However, on looking again I notice that there's a `pub is_expired()` method on the `token` that we can call instead of manually checking, so I've moved to that.